client/src/main/scala/org/http4s/client/PoolManager.scala

```diff
@@ -19,12 +23,46 @@ private final class PoolManager[F[_], A <: Connection[F]](
   private[this] val logger = getLogger(classOf[PoolManager[F, A]])
 
   private var isClosed = false
-  private var allocated = 0
-  private val idleQueue = new mutable.Queue[A]
-  private val waitQueue = new mutable.Queue[Waiting]
+  private var curTotal = 0
```

Comment:
Should this be an AtomicInteger? I don't believe var is multithread safe.

Reply:
Nope, it's only ever accessed in a synchronised block.

Reply:
If one thread calls release the code tries to decrement and one thread calls barrow the code tries to increment, I don't see where it is protected by a sync block.

Reply:
Admittedly I could just be missing it.

Reply:
So how synchronized works is that you need to acquire the lock that it's using to proceed. We are locking on the PoolManager object. So any synchronized block needs to wait if any other synchronized block currently holds the lock (not done executing). So in your case the threads would be serialized.

Reply:
I missed the fact that there is a sync block in the release call.   It seems like all the public methods have a sync block.

Reply:
It's gotten increasingly difficult to see the synchronization points with all the complexity in managing the data store.  I think we're in a good place for this feature, but there is certainly room for refactoring here.  Especially as we begin to think about other features, like max wait times, or better algorithms for connection reuse.

But I think this is good enough we need to get it into people's hands.  We'll get this into -M2, and do a quick -M3 if we missed something.